### PR TITLE
T util fix

### PR DESCRIPTION
**Pull request by Moliholy**
- Fixes many coverity issues in t_util mainly, but also others.
- Creates a UniquePtr that support _void*_ to release memory allocated via malloc/calloc

### CODE REVIEW

**Comment by jblomer** [comment]
How does this PR relates to the previous ones?  Looks like it's built #1488?

**Comment by Moliholy** [comment]
Yes, this branch keeps working on the coverity fixes, but they are both independent ones. I'll fix that

**Comment by Moliholy** [comment]
Now they should be totally independent PRs

File: cvmfs/util.h [hunk 1]
```diff
@@ -279,32 +279,66 @@ struct hash_murmur {
...
+ protected:
+  void Free() {
+    static_cast<DerivedT*>(this)->Free();
+  }
   T *ref_;
 };
 
+template <class T>
+class UniquePtr : public UniquePtrBase<T, UniquePtr<T> > {
+ private:
```
Comment (jblomer) [inline cvmfs/util.h]: Order should be `public`, `protected`, `private`

File: cvmfs/util.h [hunk 2]
```diff
@@ -279,32 +279,66 @@ struct hash_murmur {
...
+  typedef UniquePtrBase<T, UniquePtr<T> > BaseT;
+ public:
+  friend class UniquePtrBase<T, UniquePtr<T> >;
```
Comment (jblomer) [inline cvmfs/util.h]: `friend class` declarations come before `public`

File: cvmfs/util.h [hunk 3]
```diff
@@ -279,32 +279,66 @@ struct hash_murmur {
...
+  using BaseT::operator=;
```
Comment (jblomer) [inline cvmfs/util.h]: Why is this necessary?  The assignment operator is already public in the base class.

File: cvmfs/util.h (hunk 1 above)
Comment (Moliholy) [inline cvmfs/util.h]: I know, but it has to be defined in advance concretely in this case. Otherwise the _BaseT::_ three lines below won't work

File: test/unittests/t_util.cc [hunk 4]
```diff
@@ -424,7 +425,8 @@ TEST_F(T_Util, MakePipe) {
   void *buffer_output = scalloc(100, sizeof(char));
   MakePipe(fd);
   write(fd[1], to_write.c_str(), to_write.length());
-  read(fd[0], buffer_output, to_write.length());
+  int bytes_read = read(fd[0], buffer_output, to_write.length());
+  EXPECT_EQ(bytes_read, to_write.length());
```
Comment (jblomer) [inline test/unittests/t_util.cc]: This compares signed and unsigned integers, I think.

File: test/unittests/t_util.cc [hunk 5]
```diff
@@ -437,7 +439,8 @@ TEST_F(T_Util, WritePipe) {
   void *buffer_output = scalloc(20, sizeof(char));
   MakePipe(fd);
   WritePipe(fd[1], to_write.c_str(), to_write.length());
-  read(fd[0], buffer_output, to_write.length());
+  int bytes_read = read(fd[0], buffer_output, to_write.length());
+  EXPECT_EQ(bytes_read, to_write.length());
```
Comment (jblomer) [inline test/unittests/t_util.cc]: Same here

File: cvmfs/util.h (hunk 3 above)
Comment (Moliholy) [inline cvmfs/util.h]: Concretely the _operator=_ is by default both overwritten and inherited. The first one implies that the second one is _hidden_, so that _ussing BaseT::operator=_ tells the compiler not to try to create another operator= for the derived class which would hide the inherited one.

Comment (jblomer) [inline cvmfs/util.h]: I see, thanks.

File: test/unittests/t_util.cc (hunk 4 above)
Comment (Moliholy) [inline test/unittests/t_util.cc]: read() returns a ssize_t ---> __darwin_ssize_t ---> long, at least on mac

Comment (jblomer) [inline test/unittests/t_util.cc]: Yep, that's the signed one.  And `to_write.length()` is unsigned, isn't it?

File: test/unittests/t_util.cc (hunk 5 above)
Comment (Moliholy) [inline test/unittests/t_util.cc]: I think it's a long

File: test/unittests/t_util.cc (hunk 4 above)
Comment (Moliholy) [inline test/unittests/t_util.cc]: Ah sorry, you mean in the comparison. Yes, string::length() returns an unsigned int. I'll fix that with a static_cast

Comment (Moliholy) [inline test/unittests/t_util.cc]: Yes, you are right. I think it should now be fixed. In any case for the tests we run there shouldn't be a problem because we deal with very small values, always positive.

File: test/unittests/t_util.cc [hunk 6]
```diff
@@ -437,7 +439,8 @@ TEST_F(T_Util, WritePipe) {
   void *buffer_output = scalloc(20, sizeof(char));
   MakePipe(fd);
   WritePipe(fd[1], to_write.c_str(), to_write.length());
-  read(fd[0], buffer_output, to_write.length());
+  ssize_t bytes_read = read(fd[0], buffer_output, to_write.length());
+  EXPECT_EQ(static_cast<size_t>(bytes_read), to_write.length());
```
Comment (jblomer) [inline test/unittests/t_util.cc]: There are a couple of more spots in this PR where `EXPECT_EQ` compares signed and unsigned ints.

Comment (Moliholy) [inline test/unittests/t_util.cc]: Should be solved now
